pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,792 @@
+			T::InterestAccrual::convert_additive_rate_to_per_sec(rate_per_year)
+		}
+
+		fn find_write_off_state(
+			pool_id: PoolIdOf<T>,
+			maturity_date: Moment,
+		) -> Result<WriteOffState<T::Rate>, DispatchError> {
+			WriteOffState::find_best(
+				WriteOffPolicy::<T>::get(pool_id).into_iter(),
+				maturity_date,
+				T::Time::now().as_secs(),
+			)
+			.ok_or_else(|| Error::<T>::NoValidWriteOffState.into())
+		}
+
+		fn update_portfolio_valuation_with_pv(
+			pool_id: PoolIdOf<T>,
+			portfolio: &mut PortfolioValuation<T::Balance>,
+			old_pv: T::Balance,
+			new_pv: T::Balance,
+		) -> DispatchResult {
+			let prev_value = portfolio.value();
+
+			portfolio.update_with_pv_diff(old_pv, new_pv)?;
+
+			if prev_value != portfolio.value() {
+				Self::deposit_event(Event::<T>::PortfolioValuationUpdated {
+					pool_id,
+					valuation: portfolio.value(),
+					update_type: PortfolioValuationUpdateType::Inexact,
+				});
+			}
+
+			Ok(())
+		}
+
+		fn portfolio_valuation_for_pool(pool_id: PoolIdOf<T>) -> Result<T::Balance, DispatchError> {
+			let rates = T::InterestAccrual::rates();
+			ActiveLoans::<T>::get(pool_id).into_iter().try_fold(
+				T::Balance::zero(),
+				|sum, (loan, _)| -> Result<T::Balance, DispatchError> {
+					Ok(sum.ensure_add(loan.current_present_value(&rates)?)?)
+				},
+			)
+		}
+
+		fn make_active_loan<F, R>(
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			created_loan: CreatedLoan<T>,
+			f: F,
+		) -> Result<R, DispatchError>
+		where
+			F: FnOnce(&mut ActiveLoan<T>) -> Result<R, DispatchError>,
+		{
+			LatestPortfolioValuations::<T>::try_mutate(pool_id, |portfolio| {
+				ActiveLoans::<T>::try_mutate(pool_id, |active_loans| {
+					let mut loan = ActiveLoan::new(
+						loan_id,
+						created_loan.info,
+						created_loan.borrower,
+						Self::now(),
+					)?;
+
+					let result = f(&mut loan);
+
+					let last_updated = Self::now();
+					let new_pv = loan.present_value_at(last_updated)?;
+
+					Self::update_portfolio_valuation_with_pv(
+						pool_id,
+						portfolio,
+						Zero::zero(),
+						new_pv,
+					)?;
+
+					active_loans
+						.try_push((loan, last_updated))
+						.map_err(|_| Error::<T>::MaxActiveLoansReached)?;
+
+					result
+				})
+			})
+		}
+
+		fn update_active_loan<F, R>(
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			f: F,
+		) -> Result<R, DispatchError>
+		where
+			F: FnOnce(&mut ActiveLoan<T>) -> Result<R, DispatchError>,
+		{
+			LatestPortfolioValuations::<T>::try_mutate(pool_id, |portfolio| {
+				ActiveLoans::<T>::try_mutate(pool_id, |active_loans| {
+					let (loan, last_updated) = active_loans
+						.iter_mut()
+						.find(|(loan, _)| loan.loan_id() == loan_id)
+						.ok_or_else(|| {
+							if CreatedLoans::<T>::contains_key(pool_id, loan_id) {
+								Error::<T>::LoanNotActive
+							} else {
+								Error::<T>::LoanNotFound
+							}
+						})?;
+
+					*last_updated = (*last_updated).max(portfolio.last_updated());
```

Comment:
If the portfolio is updated by loan 1, then we have calculated too the PV for loan 2 in that process at `portfolio.last_updated()`.

So, if after `t` we borrow loan 2, we need to get how the PV has changed since `portfolio.last_updated()`

Reply:
You're right, good catch! I think we need to keep it two times. Last time it was computed exactly and last time it was computed inexactly. Here we will use the maximum with the *exact* time

Reply:
Taking a look at it, I think finally there is no problem.

> portfolio has been update by borrowing from another loan with id 1

This action does not change the `portfolio.last_updated()`, although now the question is if it should. It behaves the same as a legacy in this regard.

Reply:
But what does update `potfolio.last_updated()`?

Reply:
`insert_active_loan` does for example. I still think this is a bug.

Reply:
Ahh, you are right! My bad!